### PR TITLE
Optionally show templates in the service catalog

### PR DESCRIPTION
**Pull request by spadgett**
https://trello.com/c/2A7x5Pbc

Optionally adds templates to the service catalog when the template broker is not running. The template dialog itself is handled in the web console by https://github.com/openshift/origin-web-console/pull/1492

### CODE REVIEW

**Comment by spadgett** [comment]
@dtaylor113 I've added tags to `IServiceItem`

File: app/pages/home/homePageController.ts [hunk 1]
```diff
@@ -25,20 +25,12 @@ export class HomePageController {
...
   };
-
-  public navToProject = (project: any) => {
-    this.$state.go('projects/' + project.metadata.name);
-  };
-
-  public navToProjects = () => {
-    this.$state.go('projects');
-  };
 }
```
Comment (dtaylor113) [inline app/pages/home/homePageController.ts]: Was navToProject and navToProjects not used?

Comment (spadgett) [inline app/pages/home/homePageController.ts]: No longer used

File: src/components/landing-page/landing-page.controller.ts [hunk 2]
```diff
@@ -17,14 +17,12 @@ export class LandingPageController implements angular.IController {
...
     this.$scope.$on('open-overlay-panel', (event: any, item: any) => {
-      let kind = _.get(item, 'resource.kind');
-      if (kind === 'ImageStream') {
-        this.ctrl.selectedImageStream = item;
-        this.ctrl.selectedServiceClass = null;
-      } else {
-        this.ctrl.selectedImageStream = null;
-        this.ctrl.selectedServiceClass = item;
+      if (item.resource.kind === 'Template') {
+        this.ctrl.onTemplateSelected({ template: item.resource });
```
Comment (dtaylor113) [inline src/components/landing-page/landing-page.controller.ts]: Do we need a null test to see if 'onTemplateSelected' was defined?

File: app/pages/home/homePageController.ts [hunk 3]
```diff
@@ -25,20 +25,12 @@ export class HomePageController {
   };
 
   public update() {
-    this.Catalog.getCatalogItems().then( (catalogServiceItems: any) => {
+    this.Catalog.getCatalogItems(false).then( (catalogServiceItems: any) => {
```
Comment (dtaylor113) [inline app/pages/home/homePageController.ts]: I do not see where 'this.Catalog.getCatalogItems(true)' is ever called, I assume it's done in the console?

Comment (spadgett) [inline app/pages/home/homePageController.ts]: It's called in https://github.com/openshift/origin-web-console/pull/1492

Comment (spadgett) [inline app/pages/home/homePageController.ts]: https://github.com/openshift/origin-web-console/pull/1492/files#diff-9b83bfc0c5e34ad58889288622257b70R30

File: src/components/landing-page/landing-page.controller.ts (hunk 2 above)
Comment (spadgett) [inline src/components/landing-page/landing-page.controller.ts]: It's always defined even if there is no `on-template-selected` attribute.

Comment (dtaylor113) [inline src/components/landing-page/landing-page.controller.ts]: Where is it defined as a function if no `on-template-selected` attribute is passed in?  -thanks

Comment (spadgett) [inline src/components/landing-page/landing-page.controller.ts]: It's how `&` bindings work in Angular. They will always be defined.

File: src/services/catalog.service.ts [hunk 4]
```diff
@@ -21,23 +21,30 @@ export class CatalogService {
...
+    }
+
+    return this.$q.all(promises).then((values) => {
       let serviceClasses: any = values.serviceClasses.by("metadata.name");
       let imageStreams: any = values.imageStreams.by("metadata.name");
-      return this.convertToServiceItems(serviceClasses, imageStreams);
+      let templates: any = values.templates ? values.templates.by("metadata.name") : {};
+      return this.convertToServiceItems(serviceClasses, imageStreams, templates);
     }, () => {
       this.logger.log("Error Loading Catalog Items");
```
Comment (jwforres) [inline src/services/catalog.service.ts]: I feel like we should not be bombing out if one thing fails to load but the other two succeed, seems like we could at least show what was able to load and then show a message that not everything loaded successfully.  But we can open a separate issue to address this. Note that $q.all fails out at the first failure, so you'd need a different way of handling this.

Just considering the fact that the svc cat api server is separate and could potentially be down even though the master is available.

File: src/components/landing-page/landing-page.controller.ts (hunk 2 above)
Comment (spadgett) [inline src/components/landing-page/landing-page.controller.ts]: @dtaylor113 I'll change it to use the other callback style that @jeff-phillips-18 uses for consistency

File: src/services/catalog.service.ts (hunk 4 above)
Comment (spadgett) [inline src/services/catalog.service.ts]: I'm OK with that, but I think we need to break this into separate methods. It also feels wrong to return a partial result with no indication that something failed.

Comment (spadgett) [inline src/services/catalog.service.ts]: Maybe one call for each kind of item and then one call to concat and sort them?

**Comment by jwforres** [comment]
the promise can be rejected with the data you were able to retrieve as the
argument, or as part of the rejection response in combination with more
error details if necessary

**Comment by jwforres** [comment]
or this, whatever ends up being cleaner

**Comment by spadgett** [comment]
@dtaylor113 I changed the callback to a function reference instead of an object literal

File: src/services/catalog.service.ts (hunk 4 above)
Comment (spadgett) [inline src/services/catalog.service.ts]: Opened #194 for this

**Pull request merged**